UIAPPS-72 Setup release process with `auto`

joneshf-dd opened the pull request:
<!--  🎉 Hello there 🎉! Thank you for being a part of Datadog Apps! -->

## Motivation

- We want to setup a release process similar to what the Dataviz SDK uses.
- Jira Issue: https://datadoghq.atlassian.net/browse/UIAPPS-72

<!-- - Is this a bugfix or a feature? -->

## Changes

This is almost all process and infrastructure changes.
- Updated the node version to 16.3.0 for `lerna`.
- Moved over to using `auto` to handle releases.
- Added a document explaining how the releases work now.
- Added a PR Template to make `auto` a bit easier.
- Updated the GitHub actions to run `auto` instead of the previous workflow.

## Testing

<!--  Anything that would help a reviewer (or your future self) know if the change works as expected -->

Aside from dealing with CI changes, likely the only thing we can do is merge this PR and deal with whatever else we need to follow-up on.

## Change Type

<!-- Select one -->

- [x] `patch`
- [ ] `minor`
- [ ] `major`

<!-- Change Type Docs https://intuit.github.io/auto/docs/generated/pr-body-labels -->
<!-- GITHUB_RELEASE PR BODY: canary-version -->
<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @datadog/ui-extensions-sdk@0.24.1-canary.67.3d2324f.0
  # or 
  yarn add @datadog/ui-extensions-sdk@0.24.1-canary.67.3d2324f.0
  ```
</details>
<!-- GITHUB_RELEASE PR BODY: canary-version -->

joneshf-dd committed:
UIAPPS-72 Bump Node.js version to 16.3.0

We need to be on a newer version for `lerna`.

`lerna` requires at least 14.22.0 in order to work. While we could just
bump to that, most of our other repos use 16.3.0, so we bump to that
version instead.

joneshf-dd committed:
UIAPPS-72 Setup `lerna`

In order to use `auto`, we first have to use `lerna`.

`auto` depends on having `lerna` setup and working in order to do any of
its work. We bring in and init `lerna` to use an independent versioning
scheme with `yarn`. Hopefully, this will mean that we'll be able to
actual version each package independent of the others so we don't have
to bump everything all the time.

We're also only explicitly looking in the `packages` directory. Maybe
this is not quite right and we should rely on the `"workspaces"` field
in the `package.json`, but that includes the `examples` directory and
it's not clear if we'll always remember to mark each package as
`"private": true`. In any case, we can always change it later on.

joneshf-dd committed:
UIAPPS-72 Setup `auto` for managing releases

We want to make it painless to develop in this monorepo.

One of the non-trivial issues with having a monorepo is making releases
of the packages in said monorepo. There are many tools around that can
help with that and we could also roll our own. We take a page from the
Dataviz SDK and use `auto` to manage making releases.

The way it works is laid out in the `RELEASE.md` file. We can continue
to expand on this workflow as we grow.

We also added a Pull Request template to make the workflow just a little
easier. Otherwise, it's unlikely that people will know what to add in
order to make things work properly.

joneshf-dd committed:
UIAPPS-72 Add missing plugin

This plugin doesn't come with `auto`.

joneshf-dd committed:
UIAPPS-72 Move `yarn build` to publish action

We don't need to build as part of the `prepare` script.

We copied what we did with the `@datadog/ui-extensions-sdk` package and
tried to use the `prepare` script to hook into the `publish` script.
That's not correct for a few reasons:
1. `prepare` is run as part of more than just the `publish` script.
2. `release` is the script we wanted to call from the publish action.

We update things to *hopefully* work correctly now.

wfa207 commented on .github/workflows/publish.yml:
```diff
@@ -1,18 +1,23 @@
 on:
-    release:
-        types: [created]
+    - push
```
Is there ever a case where we would want to **manually** trigger publish? If so, is there another action we can use to trigger that? Possibly a tag?

joneshf-dd replied:
I don't know when we'd want to manually trigger this `publish` workflow, but we can by listening to the [`workflow_dispatch` event](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow#configuring-a-workflow-to-run-manually). That allows the `publish` workflow to be [run manually](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow#running-a-workflow).

Do  you think we should add that now?

wfa207 replied:
Eh, if we don't have a reason to yet, then I don't think it's worth spending any time on it. I think knowing that we have a path forward is enough, in case we ever need it.

joneshf-dd replied:
Okay in that case, we can at least document the idea in `RELEASE.md`.

wfa207 reviewed (approved):
This LGTM — haven't really worked with `lerna` before to manage nested packages, but seems sensible.

I had one question about whether we want to be able to trigger GH Actions with a tag in addition to merges / pushes, but it's not blocking.

joneshf-dd committed:
Merge remote-tracking branch 'origin/master' into hardy.jones/uiapps-72/setup-release-process

joneshf-dd committed:
UIAPPS-72 Set the token for `lerna`

Turns out, `lerna` won't pick up the `NPM_TOKEN` from the environment.

We're basically lifting from what the Dataviz SDK does here, but this is
a known issue. Hopefully, this change gets us back on track.

Re: https://github.com/lerna/lerna/issues/2404

joneshf-dd committed:
UIAPPS-72 Use `npm config` instead of `.npmrc`

We can't leave uncommitted changes.

`lerna` doesn't seem to work when there are uncommitted changes. When
we're writing the `.npmrc` file in CI, there are uncommitted changes.

One option is to add the `.npmrc` file to `.gitignore`, but that means
we can never track the `.npmrc` file (and this file has valid uses
outside of `lerna`). If we can't get other things to work, this might be
our only recourse.

Another option is to try using `npm config`. This should set the config
in a `~/.npmrc` file (which should not show up as an untracked file).
Whether `lerna` will respect that is to be discovered. Guess we'll know
if this code ends up staying after a while.

joneshf-dd committed:
UIAPPS-72 Document what a manual workflow could be

We don't implement it, but at least document what one option is.

joneshf-dd commented:
Aight, had to make a few updates to get the canary releases working. Mind giving that another once-over when you get a minute?

wfa207 commented on RELEASE.md:
```diff
@@ -0,0 +1,24 @@
+When a PR is opened, [`auto`][auto] will create a "canary" release, publish it to `npm`, and report back in the PR how to use the "canary" release.
+The "canary" release can be used to test out some changes without fully merging the PR.
+
+## Manual releases
+
+There is not currently a (documented) manual release process; neither through GitHub Actions nor outside of GitHub Actions.
+If we find that we need/want the ability to manually release through GitHub Actions, we can look into using the [`workflow_dispatch` event][workflow_dispatch] to [manually run a workflow][]
```
Thanks for this!

wfa207 reviewed (approved):
Looks good!

joneshf-dd commented:
Thanks for the reviews!

hydrosquall commented:
Note from the future: currently lerna is a requirement for using auto with monorepos. I looked briefly into making it work using just Yarn 2 features instead, but didn't stick with it because Lerna worked well enough for our usecase.

https://github.com/intuit/auto/discussions/1861#discussioncomment-1004780
